jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -52,56 +51,61 @@ object XGBoost extends Serializable {
   private val logger = LogFactory.getLog("XGBoostSpark")
 
   private def fromDenseToSparseLabeledPoints(
-      denseLabeledPoints: Iterator[MLLabeledPoint],
-      missing: Float): Iterator[MLLabeledPoint] = {
+      denseLabeledPoints: Iterator[XGBLabeledPoint],
+      missing: Float): Iterator[XGBLabeledPoint] = {
     if (!missing.isNaN) {
-      denseLabeledPoints.map { case MLLabeledPoint(label, features) =>
-        val dFeatures = features.toDense
-        val indices = new mutable.ArrayBuilder.ofInt()
```

Comment:
The way I see it, `ArrayBuffer` and `ArrayBuilder` serve different purposes and have different guarantees. 

`ArrayBuffer` is a Scala version of `java.util.ArrayList` which uses an Object array to store the data. Therefore, if we go with `ArrayBuffer` we would pay a significant memory/GC cost for **each** point due boxing and storing an object array instead of a primitive one.

`ArrayBuilder` on the other hand offers specialized builders for primitive array types and, as the name suggests, is there solely for the purpose of building an array (i.e. no indexing/update operations).